Add Course List block to course archive on page creation

Imran92 opened the pull request:
Implements https://github.com/Automattic/sensei/issues/6345

### Changes proposed in this Pull Request

* Added Course List block in Course Archive page

### Testing instructions

- Make sure you have an existing installation of Sensei having this branch checked out; keep it as it is
- Create a fully new local WP site containing this branch
- Install the Sensei Plugin in the new installation and complete the onboarding
- Create multiple Courses and publish them
- Go to the Courses page from frontend
- Make sure you see courses rendered by the Course List block
- Go to wp-admin->Pages->Courses, change something in the course List block, and save.
- Make sure the change takes effect in frontend
- Now delete the Course List block from the page using the editor
- Check the Courses page from frontend again, and make sure the old archive page is being rendered

### New/Updated Hooks

`sensei_course_archive_page_template` - Filters contents for the archive page being created.

### Screenshot / Video

![Screen Shot 2023-01-07 at 4 33 05 AM](https://user-images.githubusercontent.com/6820724/211113317-5982dd0b-b762-4f41-84fc-03319430e859.png)

Imran92 committed:
Add query list block for newly created courses page

Imran92 committed:
Render the page content instead of default archive template

When the page has a course list block

Imran92 committed:
Add test for welcome api page creation

Imran92 committed:
Add tests for courses archive page rendered content

Imran92 committed:
Update course list block to not break across themes

Imran92 committed:
Fix form and input fields getting removed from block

donnapep commented:
> Create a fully new local WP site containing this branch

You can save yourself some effort by deleting the existing Courses and My Courses pages, and then run the setup wizard again. It will auto-generate the new pages.

I did a bit of testing and am dropping my feedback here so far. 🙂 The below feedback is for the Course theme:

- The course progress text is centred in list view instead of left-aligned as it normally is.

![Screenshot 2023-01-13 at 9 40 43 AM](https://user-images.githubusercontent.com/1190420/212347375-be5e0ff4-4d92-4600-93e6-35fa37b73574.jpg)
- Any idea why the title is not appearing on the archive page? It used to appear prior to this change. The page template is set to "Archive" (aka default template).
- The Settings section for the block is empty, which is weird. On the My Courses page, it shows the Order By dropdown. Is there a reason to not display it here?
- Are we logging this text?
![Screenshot 2023-01-13 at 10 01 39 AM](https://user-images.githubusercontent.com/1190420/212351311-a5e8be91-6a71-47de-a4bb-2e7181b38687.jpg)

Imran92 committed:
Merge branch 'trunk' into add/course-list-block-in-archive-page

donnapep commented:
~When viewing the Courses page on Astra, it's stuck in an infinite loop for me. Are you able to reproduce?~
Logged https://github.com/Automattic/sensei/issues/6391 for this.

donnapep committed:
Merge branch 'trunk' into add/course-list-block-in-archive-page

Commit:
Merge branch 'add/course-list-block-in-my-courses-page' into add/course-list-block-in-archive-page

Imran92 committed:
Merge branch 'add/course-list-block-in-my-courses-page' into add/course-list-block-in-archive-page

Commit:
Remove course list filter from course archive page

Commit:
Render archive page content inside archive template

Commit:
Remove archive rendering code from handler

Imran92 commented:
Thanks for the review and suggestions @donnapep ! 

> Any idea why the title is not appearing on the archive page? It used to appear prior to this change. The page template is set to "Archive" (aka default template).

> The course progress text is centred in list view instead of left-aligned as it normally is.

Fixed in https://github.com/Automattic/sensei/pull/6354/commits/c5c9e87d014cdd6a2ea7e92cd16bde8ac30841e3, hope it should appear now as before.

> The Settings section for the block is empty, which is weird. On the My Courses page, it shows the Order By dropdown. Is there a reason to not display it here?

Actually, on `My Courses` page, the Query Loop block uses its own query, so it can decide how to `Order By`, but when we change the Query Loop to use the inherited global query (which is the case in the archive pages), it automatically removes that setting because the ordering is also decided by the global query. So the setting section becomes empty.

> Are we logging this text?

I didn't see it in my local environment after the new changes. Please let me know if you see it again.

I've also updated the approach to use the template we were using before. All the filters that came with it will now work with our Query Loop block. So we won't need the Course List Filter block dropdown. So we'll not need [this](https://github.com/Automattic/sensei/pull/6379) changes.

Thank you!

donnapep commented:
@Imran92 Thanks! Here are some more test results. 🙂 

- The Course Overview link is not visible on the course archive page, but it is on the My Courses page.
- I'm not sure if this is a problem with this PR or the My Courses one, but I no longer see any filters on My Courses. The Course Filter block no longer seems to be part of that page. ~Maybe that's what we wanted?~ Actually, we need the filter there because we need _Student Courses_ to default to only show active courses. And in that case the Featured and Categories filters would also need to work OR else we'd need to hide them in the block's settings for just this page.
- The course isn't styled correctly when there's a Featured label or a course category and I'm not on page 1:

![Screenshot 2023-01-18 at 12 17 52 PM](https://user-images.githubusercontent.com/1190420/213250152-ca5fa87b-6682-45ae-84f9-9584da9bba58.jpg)
- It doesn't seem like the Paid and Free filters are working properly:

![Screenshot 2023-01-18 at 12 27 02 PM](https://user-images.githubusercontent.com/1190420/213252022-c273db7a-0687-4243-b2c6-2ed81ee4ec92.jpg)

Imran92 committed:
Merge branch 'add/course-list-block-in-my-courses-page' into add/course-list-block-in-archive-page

Imran92 committed:
Add filters back in my courses page

Imran92 committed:
Merge branch 'add/course-list-block-in-archive-page' of https://github.com/Automattic/sensei into add/course-list-block-in-archive-page

Imran92 committed:
Fix lint issue

Imran92 committed:
Add course overview

donnapep commented on templates/archive-course.php:
```diff
@@ -31,26 +31,62 @@
-	<?php if ( have_posts() ) : ?>
-
-		<?php sensei_load_template( 'loop-course.php' ); ?>
+	<?php
```
Whenever we make changes to a template we also need to bump the version at the top. People often overwrite our default templates with their own, so bumping the version is an indication that they should consider updating their custom templates. Please see [here](https://nexusp2.wordpress.com/team-handbook/development-process/pull-requests/#labelling-pull-requests) for PR labelling.

Imran92 replied:
Thanks Donna! I've updated it here https://github.com/Automattic/sensei/pull/6354/commits/850ae5f683f144cfd97e44b151584f073ae9ccf0

donnapep replied:
I tweaked this in https://github.com/Automattic/sensei/pull/6354/commits/5f210b93d3684fe6f83aa889d5ca1597ecc984e3 as it indicates the plugin version in which the template was updated.

donnapep commented on templates/archive-course.php:
```diff
@@ -31,26 +31,62 @@
+	$sensei_settings_course_page = get_post( Sensei()->settings->get( 'course_page' ) );
 
-		<p><?php esc_html_e( 'No courses found that match your selection.', 'sensei-lms' ); ?></p>
+	if (
```
Generally speaking, we should try to minimize the amount of logic that is put into templates. Ideally, it should be primarily HTML.

Could we move some of this rendering logic to a function that is called from within the template? This will make it easier for those who have overwritten the default template to update it. That way, if we tweak the function, we won't have to change the template, bump the version, and go through that whole process again.

Imran92 replied:
A good suggestion, thanks Donna, moved it here https://github.com/Automattic/sensei/pull/6354/commits/a770397c6755b3cf1e466a51ed2793e29de98000

donnapep commented on includes/admin/class-sensei-setup-wizard-pages.php:
```diff
@@ -184,6 +184,105 @@ private function get_learner_courses_page_template() {
+			 * @return {array} Blocks array.
+			 */
+			apply_filters(
+				'sensei_courses_page_template',
```
I think it may be better to match the filter with the setting name (e.g. `sensei_course_archive_page_template`. And the one for My Courses probably shouldn't mention `learner` since we got rid of that term awhile back. `sensei_my_courses_page_template` would be more clear.

Imran92 replied:
Sounds nice too, updated it here https://github.com/Automattic/sensei/pull/6354/commits/8d5fefdbcfb74a736cd3c020a714b37285ff9338, will update the My Courses one in the other pr for My Courses and will deprecate the current one :)

donnapep commented on includes/admin/class-sensei-setup-wizard-pages.php:
```diff
@@ -184,6 +184,105 @@ private function get_learner_courses_page_template() {
+							'<div class="wp-block-query wp-block-sensei-lms-course-list wp-block-sensei-lms-course-list--is-list-view">
+<!-- wp:post-template {"align":"center"} -->
+<!-- wp:group {"align":"center","style":{"spacing":{"padding":{"top":"10px","right":"10px","bottom":"10px","left":"10px"}},"border":{"width":"1px","color":"#c7c3c34f"}},"className":"aligncenter","layout":{"inherit":false}} -->
+<div class="wp-block-group aligncenter has-border-color" style="border-color:#c7c3c34f;border-width:1px;padding-top:10px;padding-right:10px;padding-bottom:10px;padding-left:10px"><!-- wp:post-featured-image {"isLink":true,"height":"324px","align":"center"} /-->
```
I don't think the border color is a valid hex value. If this was copied from other code, does this issue exist there as well?

Imran92 replied:
Thanks Donna, we can change it, but it seems it's generated from the editor
<img width="575" alt="Screen Shot 2023-01-19 at 3 31 18 AM" src="https://user-images.githubusercontent.com/6820724/213307553-777522a8-b4d8-403a-aa2c-2c244fa3c1c6.png">

donnapep replied:
Ah, OK. That's the transparency then. Don't mind me!

donnapep commented on includes/admin/class-sensei-setup-wizard-pages.php:
```diff
@@ -184,6 +184,105 @@ private function get_learner_courses_page_template() {
+							'displayLayout' => [
+								'type' => 'list',
+							],
+							'className'     => 'wp-block-sensei-lms-course-list wp-block-sensei-lms-course-list\u002d\u002dis-list-view',
```
There's Unicode characters here too.

Imran92 replied:
Thanks! Removed it here https://github.com/Automattic/sensei/pull/6354/commits/6b980060b012015e952515975481922557070a78

donnapep commented on includes/admin/class-sensei-setup-wizard-pages.php:
```diff
@@ -184,6 +184,105 @@ private function get_learner_courses_page_template() {
+							'queryId'       => 0,
+							'query'         => [
+								'postType' => 'course',
+								'perPage'  => 3,
```
I don't quite get why this is set to 3, but it seems to work regardless (i.e. uses the _Blog pages show at most value_). Should we set it to 10 just for consistency's sake with My Courses? Do we even need to set it?

Imran92 replied:
Yeah the "perPage" value is ignored when query is inherited. I've removed all the values it can function without here https://github.com/Automattic/sensei/pull/6354/commits/6b980060b012015e952515975481922557070a78. Thanks!

donnapep commented on tests/unit-tests/unsupported-theme-handlers/test-class-sensei-unsupported-theme-handler-course-archive.php:
```diff
@@ -116,6 +116,27 @@ public function testShouldSetupDummyPost() {
+		$this->assertStringNotContainsString( 'wp-block-sensei-lms-course-list', $post->post_content );
+	}
+
+	public function testCourseArchivePage_WhenRendered_DoesNotRenderQueryListBlockIfPageContentAvailable() {
```
It kinda looks like the assertion in this test case is showing the block on the page?

Imran92 replied:
Good catch! Updated the name here https://github.com/Automattic/sensei/pull/6354/commits/a1cd4d7639d5830f147ad5231d00e16cb1bce200

donnapep commented:
If I add a Course List block to the page manually (i.e. delete any existing content first), but don't select a pattern, I get a fatal error and a notice when viewing the page:

![Screenshot 2023-01-18 at 3 23 12 PM](https://user-images.githubusercontent.com/1190420/213286757-f7da173d-d0c2-469b-a1c9-aeb467ff446a.jpg)

Commit:
Fix block styles not loading after certain pages

Commit:
Move content loading codes to different function

Imran92 committed:
Bump version

Imran92 committed:
Update hook name

Imran92 committed:
Remove extra preset values from wizard page, remove unicode

Imran92 committed:
Fix test case name

Imran92 committed:
Add check for queryid

Imran92 commented:
> The Course Overview link is not visible on the course archive page, but it is on the My Courses page.

Added it here https://github.com/Automattic/sensei/pull/6354/commits/558cca96c6d85ed78d391989e5b668fa6b1c36ec.

> I'm not sure if this is a problem with this PR or the My Courses one, but I no longer see any filters on My Courses. The Course Filter block no longer seems to be part of that page. Maybe that's what we wanted? Actually, we need the filter there because we need Student Courses to default to only show active courses. And in that case the Featured and Categories filters would also need to work OR else we'd need to hide them in the block's settings for just this page.

You are absolutely right, we need the filters on My Courses page, it got removed accidentally in this PR, it's there in the My Courses PR. Fixed here https://github.com/Automattic/sensei/pull/6354/commits/b1284984631db7f40b8e97cbd00b615291f6a5a7

> The course isn't styled correctly when there's a Featured label or a course category and I'm not on page 1:

This was a super tricky issue to find, figure out and fix! Very very good finding Donna! Fixed it here https://github.com/Automattic/sensei/pull/6354/commits/abfd49e30f8894880778ed3f19b2617af54208cb

Thanks @donnapep for taking careful look and the great suggestions! Can you kindly take a look again? 🦃

Imran92 commented:
> If I add a Course List block to the page manually (i.e. delete any existing content first), but don't select a pattern, I get a fatal error and a notice when viewing the page:
> 
> ![Screenshot 2023-01-18 at 3 23 12 PM](https://user-images.githubusercontent.com/1190420/213286757-f7da173d-d0c2-469b-a1c9-aeb467ff446a.jpg)

Thanks! Added a check for it here https://github.com/Automattic/sensei/pull/6354/commits/76b37d4bab1719388cef8b719e536fca1631d7e2

Imran92 committed:
Merge branch 'add/course-list-block-in-my-courses-page' into add/course-list-block-in-archive-page

donnapep committed:
Update template version number

donnapep committed:
Merge branch 'trunk' into add/course-list-block-in-archive-page

donnapep commented:
> Actually, on My Courses page, the Query Loop block uses its own query, so it can decide how to Order By, but when we change the Query Loop to use the inherited global query (which is the case in the archive pages), it automatically removes that setting because the ordering is also decided by the global query. So the setting section becomes empty.

So would you say this is more an issue with the Query Loop block? Or is it caused by us because I think we might hide other settings here via CSS? I wonder if we should consider hiding the Settings if it's empty in a separate PR. WDYT?

> I've also updated the approach to use the template we were using before. All the filters that came with it will now work with our Query Loop block. So we won't need the Course List Filter block dropdown. So we'll not need https://github.com/Automattic/sensei/pull/6379 changes.

This seems fine for now. We'll likely need to consider updating it to work with the Course List Filter block in the future. I expect we'll get questions in the call for testing though.

I noticed that the filters are above the title. Were they below the title in older versions of Sensei? I can't quite recall 🤔 .

![Screenshot 2023-01-19 at 12 03 07 PM](https://user-images.githubusercontent.com/1190420/213511495-d54664c5-78dd-4d44-8bd7-ab5f608237b0.jpg)

donnapep commented:
An FYI, that the Paid and Free filters are still not working for me. You need to enable Sensei Pro to see them.

Imran92 commented:
> So would you say this is more an issue with the Query Loop block? Or is it caused by us because I think we might hide other settings here via CSS? I wonder if we should consider hiding the Settings if it's empty in a separate PR. WDYT?

Yap absolutely we can do that in another PR, would be nice.

> This seems fine for now. We'll likely need to consider updating it to work with the Course List Filter block in the future. I expect we'll get questions in the call for testing though.

Yap I agree, it should work everywhere.

> I noticed that the filters are above the title. Were they below the title in older versions of Sensei? I can't quite recall

Hahah 😄 Had to quickly check. They were above title before too.

> An FYI, that the Paid and Free filters are still not working for me. You need to enable Sensei Pro to see them.

This is weird, they're working for me somehow, I have pro enabled. Can you kindly remove the contents of your Courses page (or switch to trunk) so that it goes back to the old archive way and see if the filters work for you then? Thanks!

donnapep commented:
> This is weird, they're working for me somehow, I have pro enabled. Can you kindly remove the contents of your Courses page so that it goes back to the old archive way and see if the filters work for you then? Thanks!

I've just tried it on the production version and get the same results so looks like it's either an existing bug or maybe corrupt data.

donnapep commented on includes/class-sensei-course.php:
```diff
@@ -4043,7 +4043,21 @@ public function get_average_days_to_completion() {
+	 * @since $$next-version$$
+	 * @return bool
+	 */
+	public function is_course_archive_page_has_content() {
```
I find this function name to be a bit misleading, as it actually checks if the page has the query block, not if it has any 'ol content. What do you think about renaming it to something like `course_archive_page_has_query_block` and updating the description accordingly? I don't think we need the `is` because there's already a `has`. 🙂

donnapep replied:
BTW, I think this is the last thing. 😉

Imran92 replied:
Hahah 😄 Thanks! Updated here https://github.com/Automattic/sensei/pull/6354/commits/37ff7776e3351a06f61f7cfbcf8c1aba5b804e0f

Imran92 committed:
Update name of course_archive_page_has_query_block function

donnapep reviewed (approved):
I've found a new issue, but I'll approve this PR and we can handle it separately.
